formatter: print request header without query string

tsaarni opened the pull request:
Signed-off-by: Tero Saarni <tero.saarni@est.tech>

**Commit Message:**
Adds formatter extension implementing a command operator that writes HTTP request header to access log without including the query string. Resolves #7583.

**Additional Description:**
Motivation for this change: Query string may contain sensitive information introducing potential [vulnerability](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url) since access logs may unintentionally expose this information. If the problem cannot be fixed in the proxied application itself, the vulnerability can be mitigated by removing query string from the access logs.

This PR is similar to #7847 but implemented as extension and it is first in-tree formatter extension (xref #14512). 

It implements command `%REQ_WITHOUT_QUERY(X?Y):Z%` which works in the same way as existing command operator `%REQ(X?Y):Z%`. The proposed implementation closely follows `SubstitutionFormatParser` and reuses `SubstitutionFormatParser::parseCommandHeader()` for parsing the command arguments.

In order to get `check_format.py` to pass, I needed to add entry to `CODEOWNERS` but since I don't know who should be added, I added @dio and myself as a placeholder for now.

**Risk Level:** Low
**Testing:** unit test
**Docs Changes:** New extension category added to docs
**Release Notes:** formatter: command operator to log request headers without query string.
**Platform Specific Features:** N/A

tsaarni commented on test/extensions/formatter/req_without_query/req_without_query_test.cc:
```diff
@@ -0,0 +1,104 @@
+  formatters:
+    - name: envoy.formatter.req_without_query
+      typed_config:
+        "@type": type.googleapis.com/google.protobuf.StringValue
```
I did not find a way to avoid requiring user to add a placeholder config like empty `StringValue`.  Is that unavoidable, since the extension point for writing formatter commands `CommandParserFactory` inherits from `Config::TypedFactory` ([link](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/include/envoy/formatter/substitution_formatter.h#L117)) instead of `UntypedFactory`?

rgs1 replied:
we need to add a config message, even if empty (e.g.: `message RequestWithoutQuery`)

tsaarni replied:
I added empty `ReqWithoutQuery` message and `#extension-category: envoy.formatter` to `ExtensionFilter` to get document generation working.

rgs1 commented on source/extensions/formatter/req_without_query/req_without_query.cc:
```diff
@@ -0,0 +1,90 @@
+
+void stripQueryString(std::string& path) {
+  const size_t query_pos = path.find('?');
+  path = std::string(path.data(), query_pos != path.npos ? query_pos : path.size());
```
Let's use Http::Utility::Url, e.g.:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/csrf/csrf_filter.cc#L40

instead of manually parsing...

tsaarni replied:
I checked `Http::Utility::Url` but I did not find a method for returning path without query string.  There seemed to be very few methods there in the [class](https://github.com/envoyproxy/envoy/blob/7d426641e044491ed13b0b45afb25604c323fbc9/source/common/http/utility.h#L135) in general. On the other hand, there seems to be several stand-alone utility functions for manipulating URLs as plain string, like [Utility::parseAndDecodeQueryString()](https://github.com/envoyproxy/envoy/blob/7d426641e044491ed13b0b45afb25604c323fbc9/source/common/http/utility.cc#L331) that seem to mostly use `url.find('?')` to separate the query string from path.

Did you mean that I should create method new method in `Http::Utility::Url` class for returning path without query string?  Or alternatively maybe move this simple `stripQueryString()` to the collection of utility functions `common/http/utility.cc`?

dio replied:
+1 for `stripQueryString()` in `Http::Utility::Url`.

tsaarni replied:
I can do that but `Http::Utility::Url` parser seems [relatively complicated](https://github.com/envoyproxy/envoy/blob/118f66e87bae3961c5726ad46239a85cdadc316f/source/common/http/utility.cc#L256), even to the extent that it uses nodejs http-parser library.  On the other hand there were other utility functions in `Http::Utility` namespace that did simple string searches for path, like [Utility::findQueryStringStart()](https://github.com/envoyproxy/envoy/blob/118f66e87bae3961c5726ad46239a85cdadc316f/source/common/http/utility.cc#L373).  I wonder if it would be OK to follow the latter example, so I've added `Utility::stripQueryString()` to this PR for now but if you prefer `Http::Utility::Url` I'll dig into nodejs http-parser library.

rgs1 reviewed (changes requested):
Generally LGTM, left two comments plus let's get the build fixed and I'll make another pass.

tsaarni commented:
> FWIW, we've achieved the equivalent of `%REQ_WITHOUT_QUERY(X?Y):Z%` by using the [header-to-metadata](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/header_to_metadata_filter) filter, e.g.:

Thanks! I did not know about this filter.

In general, I have some doubts about about this PR. Adding a complete extension for the single purpose of removing `?querystring` seems bit wrong.  Do people see it as problematic if we have a lot of small extensions that do really trivial things? 

I've also done some experiments with more general purpose formatter, inspired by your example with Header-To-Metadata, which allows following

```yaml
formatters:
  - name: envoy.formatter.regex_substitute
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.formatter.regex_substitute.v3.RegexSubstituteConfig
      command_name: REQ_WITHOUT_QUERY
      regex_value_rewrite:
        pattern:
          google_re2: {}
          regex: "^(/[^?]*).*$"
        substitution: "\\1"
```
but would this then be duplication of functionality?

dio commented:
@tsaarni sorry I was out for a health issue. Sorry was not able to update this earlier. But generally, the PR is good. If you resolve the conflict I'll take a look at it as soon as I can. And it is nice to have that REQ_WITHOUT_QUERY shorthand.

dio commented on test/extensions/formatter/req_without_query/req_without_query_test.cc:
```diff
@@ -0,0 +1,104 @@
+  TestUtility::loadFromYaml(yaml, config_);
+
+  auto formatter =
+      ::Envoy::Formatter::SubstitutionFormatStringUtils::fromProtoConfig(config_, context_.api());
```
Calling `SubstitutionFormatStringUtils::fromProtoConfig` directly doesn't work?

tsaarni replied:
It did not work, this code is in `Envoy::Extensions::Formatter` namespace.  I can add `using` if we want to avoid having long names?

tsaarni replied:
I need fully qualified names since I have two namespaces called `Formatter`.  There is `Envoy::Formatter` [here](https://github.com/envoyproxy/envoy/blob/58a13570f3f4dea9bad8b8fa5e1221d7ed5056de/include/envoy/formatter/substitution_formatter.h#L12) and I myself added `Envoy::Extensions::Formatter` causing the overlap.

The reason why I used `Envoy::Extensions::Formatter` was because the category for access log formatter extensions was already set `envoy.formatter` [here](https://github.com/envoyproxy/envoy/blob/58a13570f3f4dea9bad8b8fa5e1221d7ed5056de/include/envoy/formatter/substitution_formatter.h#L130).  I could have used e.g. `Envoy::Extensions::AccessLogFromatters` and `source/extensions/access_log_formatters/` as directory but I thought these should match with the chosen extension category name.

zuercher replied:
I think the namespaces are fine. Does just `Envoy::Formatter::SubstitutionFormatStringUtils::fromProtoConfig` work? (e.g. no leading `::`). It's also fine to add type aliases via`using` judiciously to make these more readable. That's especially true in the tests and the implementation files. For the public headers, consider the [advice in the style guilde](https://google.github.io/styleguide/cppguide.html#Aliases).

dio commented on source/extensions/formatter/req_without_query/req_without_query.h:
```diff
@@ -0,0 +1,46 @@
+class ReqWithoutQueryCommandParser : public ::Envoy::Formatter::CommandParser {
+public:
+  ReqWithoutQueryCommandParser() = default;
+  ::Envoy::Formatter::FormatterProviderPtr parse(const std::string& token, size_t,
```
`FormatterProviderPtr` doesn't work here?

tsaarni replied:
Same as above, it did not work...

tsaarni commented:
> @tsaarni sorry I was out for a health issue. Sorry was not able to update this earlier. But generally, the PR is good. If you resolve the conflict I'll take a look at it as soon as I can. And it is nice to have that REQ_WITHOUT_QUERY shorthand.

No problem with delays, glad to see you back!  I've now resolved the conflict.

I wonder if you have opinion about https://github.com/envoyproxy/envoy/pull/15711#issuecomment-819737323, that is: more general purpose formatter (regex manipulation) vs this one (dedicated for query string alone)?  I'm OK with anything at the end, but just wanted to hear your thoughts :-)

tsaarni commented:
Check `envoy-presubmit (check linux_x64 gcc)` is failing for me due to too long argument list (full log [here](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/73189/logs/146))

```
2021-04-24T19:45:27.1433253Z ERROR: /source/test/exe/BUILD:98:14: C++ compilation of rule '//test/exe:win32_scm_test' failed (Exit 1): gcc failed: error executing command 
...
2021-04-24T19:45:27.2269979Z gcc: fatal error: cannot execute ‘/usr/lib/gcc/x86_64-linux-gnu/9/cc1plus’: execv: Argument list too long
```

and in another case in different target (full log [here](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/72911/logs/141))

```
2021-04-22T18:21:55.0427418Z ERROR: /source/test/exe/BUILD:62:14: C++ compilation of rule '//test/exe:main_common_test' failed (Exit 1): gcc failed: error executing command 
...
2021-04-22T18:21:55.1126016Z gcc: fatal error: cannot execute ‘/usr/lib/gcc/x86_64-linux-gnu/9/cc1plus’: execv: Argument list too long
```

I saw @rgs1 mentioning this in https://github.com/envoyproxy/envoy/pull/14855#issuecomment-822533255 but there it was `config_fuzz_test_lib`.  I don't understand why it is not always the same target that fails but it seems to hit more compilation targets as just a specific lib.

tsaarni commented:
Could I kindly ask help in reviewing of this PR? I think @dio might be still unavailable, so I was hoping maybe others could be able to help and look at this also?

rgs1 commented on source/extensions/formatter/req_without_query/req_without_query.cc:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  str = str.substr(0, max_length.value());
+}
```
nit: this should probably be in an anonymous namespace

rgs1 reviewed (commented):
Left one nit, otherwise generally LGTM. My only remaining question is whether this should be included/built by default (it is right now, maybe it shouldn't be).

Over to @dio for another round. Thanks!

tsaarni reviewed (commented):
> Left one nit, otherwise generally LGTM. My only remaining question is whether this should be included/built by default (it is right now, maybe it shouldn't be).
> 
> Over to @dio for another round. Thanks!

Thanks @rgs1!

I guess @dio might still be away and unable to help with reviews. @snowp would you be able to help me with with getting this PR reviewed and merged?

zuercher reviewed (commented):
Hi @tsaarni -- I'm going to take over for @~dio.

Generally, this looks good. I have a couple minor comments, and I think you'll need to merge master for a conflict. Hopefully that'll fix the other CI errors. Once that's updated we'll need to pull in an API reviewer.

tsaarni commented:
> Hi @tsaarni -- I'm going to take over for @~dio.
> 
> Generally, this looks good. I have a couple minor comments, and I think you'll need to merge master for a conflict. Hopefully that'll fix the other CI errors. Once that's updated we'll need to pull in an API reviewer.

Thank you @zuercher!

I seem to have some problems with the CI since macos and windows jobs take more time than expected and get cancelled, therefore failing status.  I'm not sure if this is just a general problem since windows job finished after retrying, but macos still fails...

tsaarni commented on api/envoy/extensions/formatter/req_without_query/v3/req_without_query.proto:
```diff
@@ -0,0 +1,29 @@
+//   the HTTP request header named X first and if it's not set, then request header Y is used. If
+//   none of the headers are present '-' symbol will be in the log.
+
+// [#not-implemented-hide:]
```
Fixed.

tsaarni replied:
Fixed `Envoy::Formatter::SubstitutionFormatStringUtils::fromProtoConfig` as well i.e. removed leading `::`.

zuercher commented:
I think the macos failure is systemic. cc @envoyproxy/api-shepherds for the empty config proto

zuercher commented:
I guess @markdroth is our shepherd.

markdroth commented:
/lgtm api

zuercher commented:
@rgs1 do you want to give this a pass?

rgs1 commented:
> @rgs1 do you want to give this a pass?

Yeah I'll take another look today.

rgs1 reviewed (approved):
LGTM, thanks.

mattklein123 reviewed (changes requested):
This should have a release note. Can you add one? Thank you.

/wait

tsaarni commented:
I realised I have one question left: what `security_posture` should be set for the extension? I've set "unknown" which results in following description in the docs "_This extension has an unknown security posture and should only be used in deployments where both the downstream and upstream are trusted_". Is this ok or should the security posture be evaluated somehow more in detail?

mattklein123 commented:
> Is this ok or should the security posture be evaluated somehow more in detail?

I think it's OK to make this one trusted for downstream and upstream.

/wait

tsaarni commented:
> I think it's OK to make this one trusted for downstream and upstream.
 
Thanks @mattklein123. I've now added entry to release notes and changed the security posture.

mattklein123 commented:
Oops looks like it needs a main merge.

/wait

tsaarni commented:
/retest

tsaarni commented:
/retest

tsaarni commented:
There has been quite many flakes but these tests now seem to be failing quite consistently and I don't seem to get clean CI run anymore

* IpHttpVersions/HttpHealthCheckIntegrationTest.SingleEndpointImmediateHealthcheckFailHttp/IPv4_Http2Upstream ([logs](https://dev.azure.com/cncf/envoy/_build/results?buildId=78931&view=logs&j=110f9f1f-b6b9-58c8-998b-c13d0a153720&t=5edaee6b-1fd6-5568-80c4-e37b124f1cef&l=15632))
* IpHttpVersions/HttpHealthCheckIntegrationTest.SingleEndpointImmediateHealthcheckFailHttp/IPv6_Http2Upstream ([logs](https://dev.azure.com/cncf/envoy/_build/results?buildId=78931&view=logs&j=110f9f1f-b6b9-58c8-998b-c13d0a153720&t=5edaee6b-1fd6-5568-80c4-e37b124f1cef&l=15666))

I saw some other PRs have had same failures, so I wonder if this is a flake as well, or could it have something to do with the changes in this PR?

zuercher commented:
Yeah, it's happening on master as well.
